### PR TITLE
Video duration

### PR DESCRIPTION
**Pull request by tommasobertoni**
Homepage shows video length along with the thumbnail (issue #64):
- DurationLabel and HasDuration properties in Show model
- YouTubeShowsService downloads playlistitems for playlist info and then uses the videos id to download the videos _content details_
- _YouTubeShow displays video duration (<span> over thumbnail)
- css classes: .thumbnail-container, .video-duration

### CODE REVIEW

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 1]
```diff
@@ -137,6 +161,32 @@ private static string GetPlaylistUrl(string playlistId)
...
         }
 
+        private static string ParseDuration(string duration)
+        {
+            duration = duration.ToUpper();
+
+			// youtube api format: ISO 8601 duration
+			// https://developers.google.com/youtube/v3/docs/videos#contentDetails.duration
+            // we're interested in the sequence T(#H)#M#S (e.g. PT15M33S) but a more complicate format can be returned (e.g. P3W3DT20H31M21S)
+            var time = Regex.Match(duration, @"T(\d+H)?\d+M\d+S");
```
Comment (Eilon) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: _Strongly_ recommend not using a regex for this. It doesn't seem like it'd be too hard to parse this using some plain old regular looking code..

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 2]
```diff
@@ -22,6 +25,8 @@ public class YouTubeShowsService : IShowsService
     {
         public const string CacheKey = nameof(YouTubeShowsService);
 
+        public int ResultsCount { get; } = 3 * 8;
```
Comment (Eilon) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: Make this a private `const`?

Comment (tommasobertoni) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: actually, I was wondering if we should get the _results count_ as a parameter from the GetRecordedShowsAsync method, or even better store it in `appSettings` so it wouldn't be bound to the `YouTubeShowsService` implementation.

File: src/live.asp.net/Services/YouTubeShowsService.cs (hunk 1 above)
Comment (tommasobertoni) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: I used regex because the data is stored in a specific pattern, and I thought it'd be fun :smiley:!
If I may ask, why do you strongly recommend not using a regex for this?
In any case, I can rewrite this section without problems.

Comment (Eilon) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: My view is that regex is a big hammer, and this is a little problem :smile: Plus, regexes can be dangerous. Though the data here is fairly trusted, I don't like trusting almost anything :smile:

File: src/live.asp.net/Services/YouTubeShowsService.cs (hunk 2 above)
Comment (tommasobertoni) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: now `_resultsCount` is a private const, but I would like to know what @DamianEdwards thinks about my proposal of _injecting_ the value into the `IShowsService` implementation.

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 3]
```diff
@@ -74,25 +80,44 @@ public class YouTubeShowsService : IShowsService
...
+                var ids = string.Join(",", playlistItems.Items.Select(item => item.Snippet.ResourceId.VideoId));
+                videosRequest.Id = ids;
+
+                requestStart = DateTimeOffset.UtcNow;
+                var videos = await videosRequest.ExecuteAsync();
+                _telemetry.TrackDependency("YouTube.VideosApi", "List", requestStart, DateTimeOffset.UtcNow - requestStart, true);
+                
+                var videosContentDetailsDictionary = new Dictionary<string, Video>(videos.Items.Count);
+                foreach (var item in videos.Items)
+                    videosContentDetailsDictionary.Add(item.Id, item);
```
Comment (DamianEdwards) [inline src/live.asp.net/Services/YouTubeShowsService.cs:102]: Add braces

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 4]
```diff
@@ -74,25 +80,44 @@ public class YouTubeShowsService : IShowsService
...
+                    var videoContentDetails = videosContentDetailsDictionary[playlistItem.Snippet.ResourceId.VideoId];
+                    return new Show
+                    {
+                        Provider = _providerName,
+                        ProviderId = videoContentDetails.Id,
+                        Title = GetUsefulBitsFromTitle(playlistItem.Snippet.Title),
+                        Description = playlistItem.Snippet.Description,
+                        ShowDate = DateTimeOffset.Parse(playlistItem.Snippet.PublishedAtRaw, null, DateTimeStyles.RoundtripKind),
+                        ThumbnailUrl = playlistItem.Snippet.Thumbnails.High.Url,
+                        Url = GetVideoUrl(videoContentDetails.Id,
```
Comment (DamianEdwards) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: Why the wrapping here? If it's <120 chars long just leave it.

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 5]
```diff
@@ -137,6 +162,25 @@ private static string GetPlaylistUrl(string playlistId)
             return $"https://www.youtube.com/playlist?list={encodedPlaylistId}";
         }
 
+        private static string ParseDuration(string duration)
+        {
+            // youtube api format: ISO 8601 duration
+            // https://developers.google.com/youtube/v3/docs/videos#contentDetails.duration
+            // NB: youtube api returns the notation which includes "weeks"
```
Comment (DamianEdwards) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: NB isn't an immediately obvious prefix. Just remove it.

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 6]
```diff
@@ -137,6 +162,25 @@ private static string GetPlaylistUrl(string playlistId)
...
+            try
+            {
+                // conversion based on the W3C XML Schema Part 2: Datatypes recommendation for duration
```
Comment (DamianEdwards) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: nit: Sentence case

File: src/live.asp.net/Services/YouTubeShowsService.cs [hunk 7]
```diff
@@ -137,6 +162,25 @@ private static string GetPlaylistUrl(string playlistId)
...
+                // http://www.w3.org/TR/xmlschema-2/#duration
+                var timeSpan = ToTimeSpan(duration);
+                var hours = timeSpan.Hours > 0 ? $"{timeSpan.Hours.ToString("00")}:" : string.Empty;
+                return $"{hours}{timeSpan.Minutes.ToString("00")}:{timeSpan.Seconds.ToString("00")}";
+
+            } catch (FormatException) // duration doesn't match the format P#Y#M#DT#H#M#S (possibly includes weeks)
```
Comment (DamianEdwards) [inline src/live.asp.net/Services/YouTubeShowsService.cs]: New line

File: src/live.asp.net/wwwroot/css/site.css [hunk 8]
```diff
@@ -144,6 +145,25 @@ footer {
     font-size: 0.6em;
 }
 
+.thumbnail-container {
+    position: relative;
+}
+
+.video-duration {
+	position: absolute;
```
Comment (DamianEdwards) [inline src/live.asp.net/wwwroot/css/site.css]: Fix indent

File: src/live.asp.net/Models/Show.cs [hunk 9]
```diff
@@ -28,5 +28,9 @@ public class Show
         public string Url { get; set; }
 
         public string ThumbnailUrl { get; set; }
+
+        public string DurationLabel { get; set; }
```
Comment (DamianEdwards) [inline src/live.asp.net/Models/Show.cs]: Just call it `Duration`

**Comment by DamianEdwards** [comment]
Can you add a screenshot of what it looks like please?

**Comment by tommasobertoni** [comment]
_screenshots from Microsoft Edge_

edgelarge

_resized_

edgethin

**Comment by hishamco** [comment]
:+1:  but it will be nice if we remove black borders from the original image and move the time to the right bottom

**Comment by tommasobertoni** [comment]
I updated my fork to the current version.
@DamianEdwards please tell me if there's more to do, considering that last time this pr didn't get merged.
Maybe this isn't a desired feature?
